solver/llbsolver/ops/build.go

```diff
@@ -134,3 +143,31 @@ func (b *buildOp) Exec(ctx context.Context, inputs []solver.Result) (outputs []s
 
 	return []solver.Result{newRes.Ref}, err
 }
+
+func (b *buildOp) buildWithFrontend(ctx context.Context, inputs []solver.Result) (outputs []solver.Result, retErr error) {
+	builderInput := inputs[b.op.Builder]
+
+	wref, ok := builderInput.Sys().(*worker.WorkerRef)
+	if !ok {
+		return nil, errors.Errorf("invalid reference for build %T", builderInput.Sys())
+	}
+	rootFS := wref.ImmutableRef
+
+	cfg := specs.ImageConfig{
+		Entrypoint: b.op.Args,
+		Env:        b.op.Env,
+		WorkingDir: b.op.Cwd,
+	}
+
+	frontendInputs := make(map[string]*pb.Definition)
+	for key, def := range b.op.Defs {
+		frontendInputs[key] = def
+	}
+
+	res, err := gateway.ExecWithFrontend(ctx, b.llbBridge, b.wi, rootFS, cfg, b.op.Attrs, frontendInputs)
+	if err != nil {
+		return nil, err
+	}
+
+	return []solver.Result{res.Ref}, nil
```

Comment:
This always returns a single ref. What about if `ExecWithFrontend` result contains multiple in a map. I think we need to define some way how these map to the output indexes based on keys. Even if this isn't supported in the client pkg with the first PR it should be supported in the proto level. Also, ref arrays make this quite problematic as you can't really predict how many items build will return. Not sure what to do about that. Do we need to auto-squash arrays, or tell frontend that it may not return array results in this case.